components/TextFields/src/MDCTextInputUnderlineView.m

```diff
@@ -0,0 +1,179 @@
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+#import "MDCTextInputUnderlineView.h"
+
+#import "MaterialPalettes.h"
+
+NSString *const MDCTextInputUnderlineColorKey = @"MDCTextInputUnderlineColorKey";
+NSString *const MDCTextInputUnderlineEnabledKey = @"MDCTextInputUnderlineEnabledKey";
+NSString *const MDCTextInputUnderlineLineHeightKey = @"MDCTextInputUnderlineLineHeightKey";
+
+static const CGFloat MDCTextInputUnderlineDefaultHeight = 1.f;
+
+// TODO: (larche): Make disabled color parameterized?
+static inline UIColor *MDCTextInputUnderlineColor() {
+  return [UIColor lightGrayColor];
+}
+
+@implementation MDCTextInputUnderlineView
+
+- (instancetype)initWithFrame:(CGRect)frame {
+  self = [super initWithFrame:frame];
+  if (self) {
+    _color = MDCTextInputUnderlineColor();
+    _enabled = YES;
+    _lineHeight = MDCTextInputUnderlineDefaultHeight;
+
+    [self commonMDCUnderlineViewInit];
+  }
+  return self;
+}
+
+- (instancetype)initWithCoder:(NSCoder *)coder {
+  self = [super initWithCoder:coder];
+  if (self) {
+    _color = [coder decodeObjectForKey:MDCTextInputUnderlineColorKey];
+    _enabled = [coder decodeBoolForKey:MDCTextInputUnderlineEnabledKey];
+    _lineHeight = (CGFloat)[coder decodeFloatForKey:MDCTextInputUnderlineLineHeightKey];
+
+    [self commonMDCUnderlineViewInit];
+  }
+  return self;
+}
+
+- (void)commonMDCUnderlineViewInit {
+  [self setContentHuggingPriority:UILayoutPriorityDefaultHigh
+                          forAxis:UILayoutConstraintAxisVertical];
+  [self setClipsToBounds:NO];
+  [self updateUnderline];
+}
+
+- (void)encodeWithCoder:(NSCoder *)coder {
+  [super encodeWithCoder:coder];
+  [coder encodeObject:self.color forKey:MDCTextInputUnderlineColorKey];
+  [coder encodeBool:self.enabled forKey:MDCTextInputUnderlineEnabledKey];
+  [coder encodeFloat:(float)self.lineHeight forKey:MDCTextInputUnderlineLineHeightKey];
+}
+
+- (instancetype)copyWithZone:(NSZone *)zone {
+  MDCTextInputUnderlineView *copy = [[[self class] alloc] initWithFrame:self.frame];
+
+  copy.color = [self.color copy];
+  copy.enabled = self.enabled;
+  copy.lineHeight = self.lineHeight;
+
+  return copy;
+}
+
+- (void)layoutSubviews {
+  [super layoutSubviews];
+  [self updateUnderlinePath];
+}
+
+- (CGSize)intrinsicContentSize {
+  return CGSizeMake(UIViewNoIntrinsicMetric, self.lineHeight);
+}
+
+- (void)updateUnderlinePath {
+  CGRect bounds = [self bounds];
+  CGMutablePathRef path = CGPathCreateMutable();
+  CGFloat offSet = 1.f / [UIScreen mainScreen].scale;
+  CGPathMoveToPoint(path, NULL, CGRectGetMinX(bounds), CGRectGetMidY(bounds) + offSet);
+  CGPathAddLineToPoint(path, NULL, CGRectGetMaxX(bounds), CGRectGetMidY(bounds) + offSet);
+
+  if (_underline) {
+    _underline.frame = bounds;
+    _underline.path = path;
+  }
+
+  if (_disabledUnderline) {
+    if (path) {
```

Comment:
I surprised that _disableUnderline checks for the existence of path while _underline does not.  Should _underline check as well?
Alternatively, can we eliminate the if(path) check and assume CGPathCreateMutable() succeeds?